File: dart/lib/src/signer.dart
/// Production Ed25519 + EIP-712 signing for AI Market Protocol v2.
library;

import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';

import 'package:crypto/crypto.dart';
import 'package:ed25519_edwards/ed25519_edwards.dart' as ed25519;
import 'package:web3dart/credentials.dart';

import 'eip712.dart';

export 'eip712.dart' show computeDebitDigest, debitTypehashHeader, escrowContractName, escrowContractVersion;

class Eip712Domain {
  final String name;
  final String version;
  final int chainId;
  final String verifyingContract;

  const Eip712Domain({
    this.name = escrowContractName,
    this.version = escrowContractVersion,
    this.chainId = 8453,
    this.verifyingContract = '0x0000000000000000000000000000000000000000',
  });

  Uint8List get domainSeparatorBytes => buildDomainSeparator(
        chainId: chainId,
        verifyingContract: verifyingContract,
      );

  String get domainSeparator => MarketSigner._bytesToHex(domainSeparatorBytes);
}

class TypedData {
  final Eip712Domain domain;
  final String primaryType;
  final Map<String, dynamic> message;

  const TypedData({
    required this.domain,
    required this.primaryType,
    required this.message,
  });
}

class MarketSigner {
  final ed25519.PrivateKey _ed25519PrivateKey;
  final String? _ethereumPrivateKeyHex;

  MarketSigner({
    required String privateKeyHex,
    String? ethereumPrivateKeyHex,
  })  : _ed25519PrivateKey = ed25519.newKeyFromSeed(_parseSeedBytes(privateKeyHex)),
        _ethereumPrivateKeyHex = ethereumPrivateKeyHex;

  ed25519.PublicKey get _publicKey => ed25519.public(_ed25519PrivateKey);

  String publicKeyBase64() =>
      base64Encode(Uint8List.fromList(_publicKey.bytes));

  String publicKeyHex() => _bytesToHex(Uint8List.fromList(_publicKey.bytes));

  String signCanonical(String canonical) {
    final sig = ed25519.sign(_ed25519PrivateKey, utf8.encode(canonical));
    return 'ed25519:${base64Encode(sig)}';
  }

  bool verify(String publicKey, String signature, String canonical) {
    if (!signature.startsWith('ed25519:')) return false;
    try {
      final pubBytes = _decodePublicKey(publicKey);
      final sigBytes = base64Decode(signature.substring(8));
      final pub = ed25519.PublicKey(pubBytes);
      return ed25519.verify(pub, utf8.encode(canonical), sigBytes);
    } catch (_) {
      return false;
    }
  }

  String signEip712TypedData(TypedData data) {
    throw UnsupportedError(
      'signEip712TypedData removed — use signDebitAuthorization() for AIMarketEscrow',
    );
  }

  String signDebitAuthorization({
    required String channelId,
    required String hub,
    required String token,
    required BigInt amount,
    required String receiptId,
    required BigInt nonce,
    required int deadline,
    int chainId = 8453,
    String verifyingContract = '0x0000000000000000000000000000000000000000',
  }) {
    final ethKey = _ethereumPrivateKeyHex;
    if (ethKey == null) {
      throw StateError(
        'ethereumPrivateKeyHex is required for EIP-712 debit signing. '
        'Pass it to MarketSigner(ethereumPrivateKeyHex: ...).',
      );
    }

    final digest = computeDebitDigest(
      channelId: channelId,
      hub: hub,
      token: token,
      amount: amount,
      receiptId: receiptId,
      nonce: nonce,
      deadline: deadline,
      chainId: chainId,
      verifyingContract: verifyingContract,
    );

    final creds = EthPrivateKey.fromHex(ethKey.startsWith('0x') ? ethKey : '0x$ethKey');
    final ecSig = creds.signToEcSignature(digest, chainId: chainId);
    final r = _bytesToHex(_bigIntToFixedBytes(ecSig.r, 32));
    final s = _bytesToHex(_bigIntToFixedBytes(ecSig.s, 32));
    final v = ecSig.v.toRadixString(16).padLeft(2, '0');
    return 'eip712:0x$r$s$v';
  }

  Map<String, String> signedHeaders({
    required String channelId,
    required String capabilityId,
    required String affiliate,
  }) {
    final canonical =
        'channel:$channelId|capability:$capabilityId|affiliate:$affiliate';
    return {
      'X-Payment-Channel': channelId,
      'X-AIMarket-Affiliate': affiliate,
      'X-Market-Signature': signCanonical(canonical),
    };
  }

  String requestId(String intent) {
    final ts = DateTime.now().toUtc().millisecondsSinceEpoch;
    final data = utf8.encode('$intent:$ts');
    return sha256.convert(data).toString().substring(0, 16);
  }

  bool verifyHubSignature(String hubPublicKey, String message, String signature) {
    if (signature.startsWith('eip712:')) return false;
    final sig = signature.startsWith('ed25519:') ? signature : 'ed25519:$signature';
    return verify(hubPublicKey, sig, message);
  }

  static ({String seedHex, String publicKeyBase64}) generateEd25519Keypair() {
    final random = Random.secure();
    final seed = Uint8List.fromList(List<int>.generate(32, (_) => random.nextInt(256)));
    final pk = ed25519.newKeyFromSeed(seed);
    final pub = ed25519.public(pk);
    return (
      seedHex: _bytesToHex(seed),
      publicKeyBase64: base64Encode(Uint8List.fromList(pub.bytes)),
    );
  }

  static ({String privateKey, String address}) generateEthereumWallet() {
    final creds = EthPrivateKey.createRandom(Random.secure());
    return (
      // web3dart returns variable-length BigInt bytes (31–33); normalize to 32.
      privateKey: _bytesToHex(_normalizePrivateKeyBytes(creds.privateKey)),
      address: creds.address.hexEip55,
    );
  }

  static Uint8List _normalizePrivateKeyBytes(Uint8List raw) {
    if (raw.length == 32) return raw;
    final out = Uint8List(32);
    if (raw.length < 32) {
      out.setRange(32 - raw.length, 32, raw);
      return out;
    }
    out.setRange(0, 32, raw.sublist(raw.length - 32));
    return out;
  }

  static ({String privateKey, String address}) generateWallet() =>
      generateEthereumWallet();

  static Uint8List _parseSeedBytes(String hexOrDev) {
    final normalized = hexOrDev.startsWith('0x') ? hexOrDev.substring(2) : hexOrDev;
    if (normalized.length == 64 && RegExp(r'^[0-9a-fA-F]+$').hasMatch(normalized)) {
      return Uint8List.fromList(List<int>.generate(32, (i) {
        return int.parse(normalized.substring(i * 2, i * 2 + 2), radix: 16);
      }));
    }
    return Uint8List.fromList(sha256.convert(utf8.encode(hexOrDev)).bytes);
  }

  static Uint8List _decodePublicKey(String publicKey) {
    final trimmed = publicKey.trim();
    if (trimmed.length == 64 && RegExp(r'^[0-9a-fA-F]+$').hasMatch(trimmed)) {
      return Uint8List.fromList(List<int>.generate(32, (i) {
        return int.parse(trimmed.substring(i * 2, i * 2 + 2), radix: 16);
      }));
    }
    return Uint8List.fromList(base64Decode(trimmed));
  }

  static Uint8List _bigIntToFixedBytes(BigInt value, int length) {
    final hex = value.toRadixString(16).padLeft(length * 2, '0');
    return Uint8List.fromList(List<int>.generate(length, (i) {
      return int.parse(hex.substring(i * 2, i * 2 + 2), radix: 16);
    }));
  }

  static String _bytesToHex(Uint8List bytes) {
    return bytes.map((b) => b.toRadixString(16).padLeft(2, '0')).join();
  }
}

File: typescript/src/signer.ts
/**
 * Production Ed25519 + EIP-712 signing for AI Market Protocol v2.
 *
 * - Canonical hub messages: real Ed25519 (`ed25519:<base64>`)
 * - On-chain debit auth: keccak256 EIP-712 + secp256k1 (`eip712:0x<r><s><v>`)
 */

import nacl from 'tweetnacl';
import naclUtil from 'tweetnacl-util';
import { createHash } from 'crypto';
import { secp256k1 } from '@noble/curves/secp256k1.js';
import { type Address, type Hex, hexToBytes } from 'viem';
import { generatePrivateKey, privateKeyToAccount } from 'viem/accounts';
import { computeDebitDigest } from './eip712';

// ── EIP-712 constants ───────────────────────────────────────────────────────

export interface Eip712Domain {
  name: string;
  version: string;
  chainId: number;
  verifyingContract: string;
}

export interface TypedData {
  domain: Eip712Domain;
  primaryType: string;
  message: Record<string, string | number | bigint>;
}

export const DEBIT_TYPEHASH_HEADER =
  'DebitAuthorization(bytes32 channelId,address hub,address token,uint256 amount,bytes32 receiptId,uint256 nonce,uint256 deadline)';

export const ESCROW_CONTRACT_NAME = 'AIMarketEscrow';
export const ESCROW_CONTRACT_VERSION = '1';

export interface DebitAuthorizationParams {
  channelId: string;
  hub: string;
  token: string;
  amount: bigint;
  receiptId: string;
  nonce: bigint;
  deadline: number;
  chainId?: number;
  verifyingContract?: string;
}

export interface MarketSignerOptions {
  /** 32-byte Ed25519 seed (64-char hex) for canonical invoke signatures. */
  ed25519SeedHex: string;
  /** secp256k1 Ethereum private key (64-char hex) for EIP-712 debit auth. */
  ethereumPrivateKeyHex?: string;
}

// ── Helpers ───────────────────────────────────────────────────────────────────

function parseSeedHex(hex: string): Uint8Array {
  const normalized = hex.replace(/^0x/i, '');
  if (normalized.length === 64 && /^[0-9a-fA-F]+$/.test(normalized)) {
    const bytes = new Uint8Array(32);
    for (let i = 0; i < 32; i++) {
      bytes[i] = parseInt(normalized.slice(i * 2, i * 2 + 2), 16);
    }
    return bytes;
  }
  // Dev fallback: hash arbitrary short strings to a 32-byte seed.
  return createHash('sha256').update(hex, 'utf8').digest();
}

function parseEthPrivateKey(hex: string): Hex {
  const normalized = hex.replace(/^0x/i, '');
  if (normalized.length !== 64 || !/^[0-9a-fA-F]+$/.test(normalized)) {
    throw new Error('ethereumPrivateKeyHex must be a 32-byte hex string (64 chars)');
  }
  return `0x${normalized}` as Hex;
}

function decodePublicKey(publicKey: string): Uint8Array {
  const trimmed = publicKey.trim();
  if (/^[0-9a-fA-F]+$/.test(trimmed) && trimmed.length === 64) {
    const bytes = new Uint8Array(32);
    for (let i = 0; i < 32; i++) {
      bytes[i] = parseInt(trimmed.slice(i * 2, i * 2 + 2), 16);
    }
    return bytes;
  }
  return naclUtil.decodeBase64(trimmed);
}

function decodeEd25519Signature(signature: string): Uint8Array {
  const raw = signature.startsWith('ed25519:') ? signature.slice(8) : signature;
  return naclUtil.decodeBase64(raw);
}

// ── Signer ──────────────────────────────────────────────────────────────────

export class MarketSigner {
  private readonly seed: Uint8Array;
  private readonly keyPair: nacl.SignKeyPair;
  private readonly ethereumPrivateKeyHex?: Hex;

  constructor(seedOrOptions: string | MarketSignerOptions, ethereumPrivateKeyHex?: string) {
    const opts: MarketSignerOptions =
      typeof seedOrOptions === 'string'
        ? { ed25519SeedHex: seedOrOptions, ethereumPrivateKeyHex }
        : seedOrOptions;

    this.seed = parseSeedHex(opts.ed25519SeedHex);
    this.keyPair = nacl.sign.keyPair.fromSeed(this.seed);
    this.ethereumPrivateKeyHex = opts.ethereumPrivateKeyHex
      ? parseEthPrivateKey(opts.ethereumPrivateKeyHex)
      : undefined;
  }

  /** Ed25519 public key as base64 (hub-compatible). */
  publicKeyBase64(): string {
    return naclUtil.encodeBase64(this.keyPair.publicKey);
  }

  /** Ed25519 public key as hex. */
  publicKeyHex(): string {
    return Buffer.from(this.keyPair.publicKey).toString('hex');
  }

  /** Sign a canonical UTF-8 string → `ed25519:<base64>`. */
  signCanonical(canonical: string): string {
    const message = new TextEncoder().encode(canonical);
    const sig = nacl.sign.detached(message, this.keyPair.secretKey);
    return `ed25519:${naclUtil.encodeBase64(sig)}`;
  }

  /** Verify `ed25519:<base64>` with hex- or base64-encoded public key. */
  verify(publicKey: string, signature: string, canonical: string): boolean {
    if (!signature.startsWith('ed25519:')) return false;
    try {
      const message = new TextEncoder().encode(canonical);
      const sig = decodeEd25519Signature(signature);
      const pub = decodePublicKey(publicKey);
      return nacl.sign.detached.verify(message, sig, pub);
    } catch {
      return false;
    }
  }

  /** Sign EIP-712 debit authorization → `eip712:0x<130 hex chars>`. */
  signDebitAuthorization(params: DebitAuthorizationParams): string {
    const ethKey = this.ethereumPrivateKeyHex;
    if (!ethKey) {
      throw new Error(
        'ethereumPrivateKeyHex is required for EIP-712 debit signing. ' +
          'Pass it to MarketSigner constructor as second argument or via MarketSignerOptions.'
      );
    }

    const digest = computeDebitDigest({
      channelId: params.channelId as Hex,
      hub: params.hub as Address,
      token: params.token as Address,
      amount: params.amount,
      receiptId: params.receiptId as Hex,
      nonce: params.nonce,
      deadline: BigInt(params.deadline),
      chainId: params.chainId ?? 8453,
      verifyingContract: (params.verifyingContract ??
        '0x0000000000000000000000000000000000000000') as Address,
    });

    const digestBytes = hexToBytes(digest);
    const privBytes = hexToBytes(ethKey);
    const recovered = secp256k1.sign(digestBytes, privBytes, {
      lowS: true,
      format: 'recovered',
    });
    const r = Buffer.from(recovered.slice(0, 32)).toString('hex');
    const s = Buffer.from(recovered.slice(32, 64)).toString('hex');
    const v = recovered[64] + 27;
    return `eip712:0x${r}${s}${v.toString(16).padStart(2, '0')}`;
  }

  signEip712TypedData(_data: TypedData): string {
    throw new Error('signEip712TypedData removed — use signDebitAuthorization() for AIMarketEscrow');
  }

  signedHeaders(args: {
    channelId: string;
    capabilityId: string;
    affiliate: string;
  }): Record<string, string> {
    const canonical = `channel:${args.channelId}|capability:${args.capabilityId}|affiliate:${args.affiliate}`;
    return {
      'X-Payment-Channel': args.channelId,
      'X-AIMarket-Affiliate': args.affiliate,
      'X-Market-Signature': this.signCanonical(canonical),
    };
  }

  verifyHubSignature(hubPublicKey: string, message: string, signature: string): boolean {
    if (signature.startsWith('eip712:')) return false;
    const canonical = signature.startsWith('ed25519:') ? message : message;
    const sig = signature.startsWith('ed25519:') ? signature : `ed25519:${signature}`;
    return this.verify(hubPublicKey, sig, canonical);
  }

  static generateEd25519Keypair(): { seedHex: string; publicKeyBase64: string } {
    const seed = nacl.randomBytes(32);
    const pair = nacl.sign.keyPair.fromSeed(seed);
    return {
      seedHex: Buffer.from(seed).toString('hex'),
      publicKeyBase64: naclUtil.encodeBase64(pair.publicKey),
    };
  }

  /** Generate a secp256k1 Ethereum wallet for channel deposits / EIP-712. */
  static generateEthereumWallet(): { privateKeyHex: Hex; address: Address } {
    const privateKey = generatePrivateKey();
    const account = privateKeyToAccount(privateKey);
    return { privateKeyHex: privateKey, address: account.address };
  }

  /** @deprecated Prefer [generateEd25519Keypair] or [generateEthereumWallet]. */
  static generateWallet(): { privateKey: string; address: string } {
    const eth = MarketSigner.generateEthereumWallet();
    return { privateKey: eth.privateKeyHex.slice(2), address: eth.address };
  }
}

// Re-export digest helper at package boundary.
export { computeDebitDigest } from './eip712';
